expyfun/stimuli/_tracker.py

```diff
@@ -0,0 +1,858 @@
+"""Adative tracks for psychophysics (individual, or multiple randomly dealt)
+"""
+# Author: Ross Maddox <ross.maddox@rochester.edu>
+#
+# License: BSD (3-clause)
+
+import numpy as np
+import time
+from scipy.stats import binom as binom
+import json
+import matplotlib.pyplot as plt
+from .. import ExperimentController
+
+
+# =============================================================================
+# Set up the logging callback (use write_data_line or do nothing)
+# =============================================================================
+def callback_dummy(event_type, value=None, timestamp=None):
+    """Take the arguments of write_data_line, but do nothing.
+    """
+    pass
+
+
+def check_callback(callback):
+    """Check to see if the callback is of an allowable type.
+    """
+    if callback is None:
+        callback = callback_dummy
+    elif isinstance(callback, ExperimentController):
+        callback = callback.write_data_line
+
+    if not callable(callback):
+        raise(TypeError,
+              'callback must be a callable, None, or an instance of '
+              'ExperimentController.')
+    return callback
+
+
+# =============================================================================
+# Define the TrackerUD Class
+# =============================================================================
+class TrackerUD(object):
+    """Up-down adaptive tracker
+
+    This class implements a standard up-down adative tracker object. Based on
+    how it is configured, it can be used to run a fixed-step m-down n-up
+    tracker (staircase), or it can implement a weighted up-down procedure.
+
+    Parameters
+    ----------
+    callback : callable | ExperimentController | None
+        The function that will be used to print the data, usually to the
+        experiment .tab file. It should follow the prototype of
+        ``ExperimentController.write_data_line``. If an instance of
+        ``ExperimentController`` is given, then it will take that object's
+        ``write_data_line`` function. If None is given, then it will not write
+        the data anywhere.
+    up : int
+        The number of wrong answers necessary to move the tracker level up.
+    down : int
+        The number of correct answers necessary to move the tracker level down.
```

Comment:
The level would equate to something like intensity or contrast. This is in line with the "good thresholds are low" frame of reference, and also results in a 3-down 1-up tracker operating as they are described in the literature.